persist-tool/add.c

```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright (c) 2019 Balabit
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "add.h"
+
+typedef struct __PersistStateEntry
+{
+  gchar *name;
+  gchar *value;
+} PersistStateEntry;
+
+FILE *input_file;
+
+gboolean
+check_directory_exists(gchar *path)
+{
+  return g_file_test(path, G_FILE_TEST_EXISTS | G_FILE_TEST_IS_DIR);
+}
+
+PersistStateEntry *
+persist_state_entry_new(void)
+{
+  PersistStateEntry *self = g_new0(PersistStateEntry, 1);
+  return self;
+}
+
+gboolean
+persist_state_parse_entry(PersistStateEntry *self, gchar *entry)
+{
+  gchar *p = strchr(entry, '=');
+  if (!p)
+    {
+      return FALSE;
+    }
+  self->name = g_strndup(entry, p - entry);
+  self->name = g_strstrip(self->name);
+  self->value = g_strdup((p+1));
+  return TRUE;
+}
+
+#define NAME_LENGTH 40
+#define MAX_LINE_LEN 4000
+#define MAX_ENTRY_VALUE_LENGTH (3*MAX_LINE_LEN)
+
+/* This function parses a JSON style string
+ * containing series of integer values.
+ * For example:
+ * { "value": "00 00 00 00 01 02 03 0A " } */
+gint
+parse_value_string(gchar *value_str, gint *buffer, gint buffer_length)
+{
+  gchar *p = strchr(value_str, ':');
+  if (!p)
+    return -1;
+
+  p = strchr(p, '"');
+  if (!p || strlen(p)<1)
+    return -1;
+
+  p = g_strchug(p+1); // p points to the first value
+
+  gint count = 0;
+  while (count < buffer_length)
+    {
+      int value;
+      int result = sscanf(p, "%X", &value);
+
+      if (result == EOF || result == 0)
```

Comment:
sscan returns with 0 or EOF not only in case of errors. The same will happen when we reach the end of the string or string contains more than one space as separator.

Reply:
Indeed. 0 and EOF are not enough for error detection, but there is an extra piece of information that can be used: the list must end with doublequote `"`. With that involved, the parse stop can be detected properly.

However, with the rewrite with `g_strsplit`, this will probably be resolved automatically: you will probably need to copy the string between the doublequotes before calling g_strsplit into a nullterminated string. So the doublquote will be automatically taken into account.